src/bsp/platform/stm32f4_nucleo_wifi/xi_bsp_io_net_stm32f4_nucleo_wifi.c

```diff
@@ -0,0 +1,235 @@
+#endif
+
+#define XI_BSP_IO_NET_BUFFER_SIZE 512
+
+typedef enum {
+    xi_wifi_state_connected = 0,
+    xi_wifi_state_disconnected,
+    xi_wifi_state_error,
+} xi_bsp_stm_wifi_state_t;
+
+typedef struct _xi_bsp_stm_wifi_buffer_t
+{
+    uint8_t bytes[XI_BSP_IO_NET_BUFFER_SIZE];
+    uint32_t size;
+} xi_bsp_stm_wifi_buffer_t;
+
+xi_bsp_stm_wifi_state_t xi_wifi_state   = xi_wifi_state_connected;
+xi_bsp_stm_wifi_buffer_t xi_wifi_buffer = {0};
+
+xi_bsp_io_net_state_t xi_bsp_io_net_create_socket( xi_bsp_socket_t* xi_socket )
+{
+    ( void )xi_socket;
+
+    *xi_socket = 0;
+
+    return XI_BSP_IO_NET_STATE_OK;
+}
+
+xi_bsp_io_net_state_t
+xi_bsp_io_net_connect( xi_bsp_socket_t* xi_socket, const char* host, uint16_t port )
+{
+    ( void )xi_socket;
+    ( void )host;
+    ( void )port;
+
+    char* protocol = "s"; // t -> tcp , s-> secure tcp, c-> secure tcp with certs
+
+    WiFi_Status_t status = WiFi_MODULE_SUCCESS;
+    status = wifi_socket_client_open( ( uint8_t* )"broker.dev.xively.io", port,
+                                      ( uint8_t* )protocol, ( uint8_t* )xi_socket );
+
+    if ( status == WiFi_MODULE_SUCCESS )
+    {
+        return XI_BSP_IO_NET_STATE_OK;
+    }
+    else
+    {
+        return XI_BSP_IO_NET_STATE_ERROR;
+    }
+}
+
+xi_bsp_io_net_state_t xi_bsp_io_net_connection_check( xi_bsp_socket_t xi_socket,
+                                                      const char* host,
+                                                      uint16_t port )
+{
+    ( void )xi_socket;
+    ( void )host;
+    ( void )port;
+
+    if ( xi_wifi_state == xi_wifi_state_connected )
+    {
+        return XI_BSP_IO_NET_STATE_OK;
+    }
+    else
+    {
+        return XI_BSP_IO_NET_STATE_ERROR;
+    }
+}
+
+xi_bsp_io_net_state_t xi_bsp_io_net_write( xi_bsp_socket_t xi_socket,
+                                           int* out_written_count,
+                                           const uint8_t* buf,
+                                           size_t count )
+{
+    ( void )xi_socket;
+    ( void )buf;
+    *out_written_count = count;
+
+    WiFi_Status_t status = WiFi_MODULE_SUCCESS;
+    status               = wifi_socket_client_write( xi_socket, count, ( char* )buf );
+
+    if ( status == WiFi_MODULE_SUCCESS )
+    {
+        return XI_BSP_IO_NET_STATE_OK;
+    }
+    else
+    {
+        return XI_BSP_IO_NET_STATE_ERROR;
+    }
+}
+
+xi_bsp_io_net_state_t xi_bsp_io_net_read( xi_bsp_socket_t xi_socket,
+                                          int* out_read_count,
+                                          uint8_t* buf,
+                                          size_t count )
+{
+    ( void )xi_socket;
+    ( void )buf;
+    ( void )count;
+
+    *out_read_count = 0;
+
+    if ( xi_wifi_state == xi_wifi_state_connected )
+    {
+        if ( xi_wifi_buffer.size == 0 )
+        {
+            return XI_BSP_IO_NET_STATE_BUSY;
+        }
+        else if ( xi_wifi_buffer.size > count )
```

Comment:
Actually I was thinking about taking from the buffer the ```min( count, xi_wifi_buffer.size )``` and then leaving in the buffer the rest - it's a little bit more complicated but should provide more flexibility.